components/scream/src/physics/shoc/shoc_ic_cases.cpp

```diff
@@ -0,0 +1,185 @@
+    }
+  }
+
+  // Move up the column, computing the pressures at cell centers and
+  // interfaces.
+  for (Int j = 1; j < nlev; ++j) {
+    Real z0 = z(i, j-1);
+    Real z1 = z(i, j);
+    Real th0 = interpolate_data(z_ref, theta_ref, z0);
+    Real th1 = interpolate_data(z_ref, theta_ref, z1);
+    Real p0 = pres(i, j-1);
+    if (std::abs(th0 - th1) < 1e-14 * th0) {
+      pres(i, j) = pow(pow(p0, k) + k*c*(z1 - z0)/th0, 1.0/k);
+    }
+    else {
+      Real ra = (z1 - z0)/(th1 - th0);
+      pres(i, j) = pow(pow(p0, k) + k*c*ra*log(th1/th0), 1.0/k);
+    }
+  }
+}
+
+void initialize_column_data(Real ztop, Int col, FortranData& d) {
+  using consts = Constants<Real>;
+  const Int i = col;
+  const Int nlev = d.nlev;
+  const Real dz = ztop/nlev;
+
+  // Set the horizontal grid spacing.
+  d.host_dx(i) = 5300.0;
+  d.host_dy(i) = 5300.0;
+
+  for (Int k = 0; k < nlev; ++k) {
+
+    // Set up the vertical grid.
+    Real zi = k * dz;
+    Real zt = (k+0.5) * dz;
+    d.zi_grid(i, k) = zi;
+    d.zt_grid(i, k) = zt;
+
+    // Interpolate the potential temperature.
+    const Real theta_zt = interpolate_data(z_ref, theta_ref, zt);
+    const Real qw = interpolate_data(z_ref, qw_ref, zt);
+    const Real ql = interpolate_data(z_ref, ql_ref, zt);
+    d.qw(i, k) = qw;
+    d.shoc_ql(i, k) = ql;
+    Real zvir = (consts::RH2O / consts::RAir) - 1.0;
+    d.thv(i, k) = theta_zt * (1.0 + zvir * qw);
+    d.thetal(i, k) = theta_zt;
+
+    // Set cell-centered wind speeds.
+    d.u_wind(i, k) = interpolate_data(wind_z_ref, u_ref, zt);
+    d.v_wind(i, k) = interpolate_data(wind_z_ref, v_ref, zt);
+    d.w_field(i, k) = interpolate_data(wind_z_ref, w_ref, zt);
+
+    // Set turbulent kinetic energy.
+    // TODO: Not done yet!
+
+    // Set tracers.
+    for (Int q = 0; q < d.num_qtracers; ++q)
+    {
+      d.qtracers(i, k, q) = sin(q/3. + 0.1 * zt * 0.2 * (k+1));
+      d.wtracer_sfc(i, q) = 0.1 * k;
+    }
+
+    // Set surface fluxes and forcings.
+    // (Not clear whether these can be set independent of mesh)
+    d.wthl_sfc[i] = 1e-4;
+    d.wqw_sfc[i] = 1e-6;
+    d.uw_sfc[i] = 1e-2;
+    d.vw_sfc[i] = 1e-4;
+  }
+
+  // Compute hydrostatic pressure at cell centers and interfaces.
+  compute_column_pressure(i, d.nlev, d.zt_grid, d.pres);
+  compute_column_pressure(i, d.nlevi, d.zi_grid, d.presi);
+
+  // Compute pressure differences.
+  for (Int k = 0; k < nlev; ++k) {
+    d.pdel(i, k) = d.presi(i, k+1) - d.presi(i, k);
+  }
+
+  // Initialize host_dse and exner, which are assumed to be valid inputs to
+  // shoc_main.
+  for (Int k = 0; k < nlev; ++k) {
+    d.exner(i, k) = pow(d.pres(i, k)/consts::P0, consts::RAir/consts::CpAir);
+    d.host_dse(i, k) = consts::CpAir * d.exner(i, k) * d.thv(i, k) +
+                       consts::GGr * d.zt_grid(i, k);
+  }
+}
+
+} // end anonymous namespace
+
+// From scream-docs/shoc-port/shocintr.py.
+FortranData::Ptr Factory::create (Int shcol, Int nlev) {
```

Comment:
Then this factory would take a default enum as input indicating which case to use.

Reply:
I notice that the p3 factory only takes the number of columns as an argument, and that the number of vertical levels seems to be 72. I know that we're still stuck with a handful of "blessed" numbers for vertical levels, but shouldn't we allow this to be specified when we create a set of initial conditions?

Reply:
Yep, good generalization.